cardano-mempool/tests/Test/Cardano/Memory/PoolTests.hs

```diff
@@ -106,85 +108,107 @@ mallocPreFilled preFillByte bc = do
   withForeignPtr mfp $ \ptr -> setPtr (castPtr ptr) bc preFillByte
   pure mfp
 
+-- | Wait for all items to be GCed by periodicaly trigger garbage collector for certain
+-- number of times before failing, unless the counter action was executed the exact number
+-- of times.
```

Comment:
I would consider rewriting this as 2 separate functions: `ensureAllGCedWith`, which takes number of iterations and delay between iterations as arguments, and then `ensureAllGCed` defined in terms of that. This would make these two parameters explicit, instead of burying them in the body of the function and mentioning them in the Haddock as "periodically" and "a certain number of times". IMO this would be better than just documenting the values, because `ensureAllGCed` would be very small (just 1 line, really), reducing the odds of "documentation rot".

I would also try to come up with more descriptive names for `n` and `f`.

So something like this:

```haskell
-- | @ensureAllGCedWith iterations delay expectedCount registerCounter@ waits
-- for all items to be GCed by triggering garbage collection @iterations@
-- times, once every @delay@ milliseconds. After @iterations@ attempts, if the
-- counter hook was not called exactly @expectedCount@ times, a test failure is
-- raised via 'assertFailure'. Garbage collection is tracked via finalizers
-- (see below).
ensureAllGCedWith ::
       Int
       -- ^ Number of GC attempts to make before failing
    -> Integer
       -- ^ Delay between attempts, in milliseconds
    -> Int
       -- ^ Expected number of counter hook firings (in practice: individual
       -- garbage collections on 'ForeignPtr's, as per their finalizers).
    -> (IO () -> IO a)
       -- ^ Function for registering the counter hook. The argument to this
       -- function should be attached to each 'ForeignPtr' we're interested in
       -- as a finalizer.
    -> IO a
ensureAllGCedWith iterations delay expectedCount registerCounter = do
  -- existing impl here

-- | 'ensureAllGCedWith' with default values: 100 iterations, 10ms delay.
ensureAllGCed :: Int -> (IO () -> IO a) -> IO a
ensureAllGCed = ensureAllGCedWith 100 10
```

Reply:
I am not sure I agree with all these comments. This is a helper function that is used within this test module only. Haddock will never be generated for this function and is here merely for the purpose of code comment, rather than documentation. Moreover I don't like having the size of comments to be more than the size of actual code. Anyone reading this function should just look at the code rather than reading lengthy documentation to see what that function is doing.

Adding an extra function adds an unnecessary layer of indirection and indicates that it can be used on its own as well, which is not the case here.

If this was an exported function that was part of the library, I would have agreed more with most of your comments.

Despite all that I will change it to your suggested version, just because with test suites I care more about correctness, rather than readability.